exploration/0477-quoted-literals.md

```diff
@@ -0,0 +1,268 @@
+  > match {$arbitraryString}
+  > when |can't resolve| {Can't resolve!}
+  > when |11'233.44| {Locale formatted number}
+  > when |New York| {A multi-word proper name}
+  > when * {Imagine more...}
+  > ```
+
+- Message authors may want to protect untranslatable strings:
+
+  > ```
+  > {Visit {|http://www.example.com| @translate=false}.}
+  > ```
+  >
+  > See [design proposal 0002](https://github.com/unicode-org/message-format-wg/blob/main/exploration/0002-expression-attributes.md).
+
+- Message authors may want to decorate substrings as being written in a particular language,
+  different from the message's language,
+  for the purpose of accessibility, text-to-speech, and semantic correctness.
+
+  > ```
+  > {The official native name of the Republic of Poland is {|Rzeczpospolita Polska| @lang=pl}.}
+  > ```
+  >
+  > See [design proposal 0002](https://github.com/unicode-org/message-format-wg/blob/main/exploration/0002-expression-attributes.md).
+
+- Developers may want to embed messages with quoted literals in code written in another programming language
+  which uses single or double quotes to delimit strings.
+
+  > ```js
+  > let message = new MessageFormat("en", "{A message with {|a literal|}.}");
+  > ```
+
+- Developers and localization engineers may want to embed messages with quoted literals in a container format, such as JSON.
+
+  > ```json
+  > {
+  >   "msg": "{A message with {|a literal|}.}"
+  > }
+  > ```
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- **[r1; high priority]** Minimize the need to escape characters inside literals.
+  In particular, choose a delimiter that isn't frequently used in translation content.
+  Having to escape characters inside literals is inconvenient and error-prone when done by hand,
+  and it also introduces the backslash into the message, `\`,
+  which is the escape introducer.
+  The backslash then needs to be escaped too,
+  when the message is embedded in code or containers.
+  (This is how some syntaxes produce the gnarly `\\\`.)
+
+- **[r2; high priority]** Minimize the need to escape characters when embedding messages in code or containers.
+  In particular, choose a delimiter that isn't frequently used as a string delimiter in programming languages and container formats.
+  However, note that many programming languages also provide alternative ways of delimiting strings, e.g. _raw strings_ or triple-quoted literals.
+
+- **[r3; medium priority]** Minimize the incentive to avoid escaping by changing messages
+  (e.g. rephrasing content, using typographic apostrophes, or switching outer delimiters).
+
+- **[r4; medium priority]** Don't surprise users with syntax that's too exotic.
+  We expect quoted literals to be rare,
+  which means fewer opportunities to get used to their syntax and remember it.
+
+- **[r5; low priority]** Be able to pair the opening and the closing delimiter,
+  to aid parsers recover from syntax errors,
```

Comment:
I don't see how paired delimiters would help parsers in recovery. A missing quote delimiter of any sort is pretty fatal for the whole message, even for a parser that can do some recovery.

If we did want to actually help here, then requiring `{}` to be escaped in quoted literals would let the damage be limited in most cases to a single expression.

Reply:
Paired delimiters don't help runtime parsers in recovery (as there is effectively no recovery in many cases), but they do help a different sort of parser: editors and other tools often auto-pair enclosing punctuation. So I'd probably phrase this differently and I might boost the priority a little.

The problem is that there are not many ASCII paired characters: `()`, `{}`, and `[]`

Reply:
> I don't see how paired delimiters would help parsers in recovery. A missing quote delimiter of any sort is pretty fatal for the whole message, even for a parser that can do some recovery.

I admit that the technique that I had in mind here is rather obscure. Namely, if both paired delimiters are made special in the literal, then the start of another literal can be an anchor point for a parser to resume parsing.

```
    There {:is a=|broken literal=|here|}
                        ^         ^
                        The closing delimiter is missing here.
                                  The syntax error occurs here.

    There {:is a=[broken literal=[here]}
                        ^^       ^ 
                        The closing delimiter is missing here.
                         |       The parser can recognize a new literal here...
                         and rewind to here.
```

I'll add a `<details>` block to explain this.